### PR TITLE
Add a new op CollectiveReduce to StableHLo

### PR DESCRIPTION
**Pull request by Tixxx**
This is following the design doc [here](https://github.com/openxla/xla/discussions/45352) to add a new op CollectiveReduce to stableHlo and xla.

### CODE REVIEW

**Comment by GleasonK** [comment]
Thanks! 

Please review vhlo_checklist.md in the shlo repo (will link once back at computer), missing some tests and required changes in there

**Comment by Tixxx** [comment]
> Thanks!
> 
> Please review vhlo_checklist.md in the shlo repo (will link once back at computer), missing some tests and required changes in there

ah ok, didnt see that, let me add missing items

**Comment by Tixxx** [comment]
@GleasonK please take another look and let me know if im missing anything.

**Comment by Tixxx** [comment]
@GleasonK  gentle ping on the review

File: docs/spec.md [hunk 1]
```diff
@@ -1916,6 +1916,101 @@ Afterwards, `result@process` is given by:
...
+
+Within each process group in the StableHLO process grid, send the value of the
+`operands` tensor to the source process, applies a reduction
+function `computation` to the values of the `operands` tensors from each process
+and produces `results` tensors and save the
+`results` tensors only on the source process.
+
+If `has_dynamic_root` is set to true, the last operand in `operands` must be
+a flat array of `si32` containing the root values with 1-to-1 mapping to each
+data operands.
```
Comment (mrguenther) [inline docs/spec.md]: Nit: "to each data operand" for singular/plural agreement

File: docs/spec.md [hunk 2]
```diff
@@ -1916,6 +1916,101 @@ Afterwards, `result@process` is given by:
...
 
+### collective_reduce
+
+#### Semantics
+
+Within each process group in the StableHLO process grid, send the value of the
+`operands` tensor to the source process, applies a reduction
+function `computation` to the values of the `operands` tensors from each process
+and produces `results` tensors and save the
+`results` tensors only on the source process.
```
Comment (mrguenther) [inline docs/spec.md:1927]: Nit: tense agreement (imperative, e.g. "save," vs. indicative, e.g. "saves")

**Review by GleasonK — COMMENTED** [review]
Reviewed changes in 
- `stablehlo/transforms/`
- `stablehlo/dialect/vhlo`

LGTM. Letting @mrguenther take a closer look at the rest of the code, but overall lgtm.

File: stablehlo/dialect/StablehloOps.cpp [hunk 3]
```diff
@@ -1844,8 +1883,8 @@ LogicalResult AsyncStartOp::inferReturnTypes(
 
   Operation* collectiveOp = &block.front();
   if (!isa<AllGatherOp, AllReduceOp, AllToAllOp, CollectiveBroadcastOp,
-           CollectivePermuteOp, ReduceScatterOp, SliceOp, DynamicSliceOp,
-           DynamicUpdateSliceOp>(collectiveOp)) {
+           CollectiveReduceOp, CollectivePermuteOp, ReduceScatterOp, SliceOp,
+           DynamicSliceOp, DynamicUpdateSliceOp>(collectiveOp)) {
```
Comment (mrguenther) [inline stablehlo/dialect/StablehloOps.cpp]: Nit: It looks like these were once alphabetized. Could we re-alphabetize them just for cleanliness?

File: stablehlo/reference/Ops.cpp [hunk 4]
```diff
@@ -1443,6 +1456,81 @@ Tensor collectiveBroadcastOp(const Tensor& operand,
...
+  SmallVector<InterpreterValue> results(resultTypes.size());
+  for (const auto& [resultIndex, resultType] : llvm::enumerate(resultTypes)) {
+    // Find root process for this operand.
+    ProcessId rootId = (*processGroup)[0];  // default: first in group
+    if (hasDynamicRoot) {
+      // Root index comes from the last operand's resultIndex-th element.
+      const Tensor& rootTensor = operands.back();
+      auto rootIdx = rootTensor.get({static_cast<int64_t>(resultIndex)});
+      rootId = (*processGroup)[static_cast<size_t>(
+          rootIdx.getIntegerValue().getSExtValue())];
```
Comment (mrguenther) [inline stablehlo/reference/Ops.cpp]: Did you mean to cast a sign-extended value to an unsigned type?

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -1443,6 +1456,81 @@ Tensor collectiveBroadcastOp(const Tensor& operand,
...
+      }
+      resultTensor.set(*elementIndex, reduced.get({}));
+    }
+
+    // Only root gets the reduced result; others get zeros.
+    if (process->getId() == rootId)
+      results[resultIndex] = resultTensor;
+    else
+      results[resultIndex] = broadcastInDimOp(
+          constant(0.0, resultType.getElementType()), {}, resultType);
```
Comment (mrguenther) [inline stablehlo/reference/Ops.cpp]: Couldn't we check this before initializing `resultTensor` and skip initializing the unused value if `process->getId() != rootId`?

**Comment by mrguenther** [comment]
@GleasonK Should this PR update `stablehlo/tests/ops_stablehlo.mlir`?

**Review by mrguenther — COMMENTED** [review]
LGTM modulo a few minor comments.

File: stablehlo/reference/Ops.cpp (hunk 5 above)
Comment (Tixxx) [reply stablehlo/reference/Ops.cpp]: good catch, updated

File: stablehlo/reference/Ops.cpp (hunk 4 above)
Comment (Tixxx) [reply stablehlo/reference/Ops.cpp]: just realized there's a getZExtValue, simplified  this

**Comment by Tixxx** [comment]
> @GleasonK Should this PR update `stablehlo/tests/ops_stablehlo.mlir`?

added some tests in `stablehlo/tests/ops_stablehlo.mlir` too, actually caught a bug in the multi-operand case.
Could you take another look? thanks

File: stablehlo/tests/ops_stablehlo.mlir [hunk 6]
```diff
@@ -6837,3 +6837,232 @@ func.func @all_gather_with_mesh_axes_replica_group(%operand: tensor<16x8xf32>) -
...
+  "stablehlo.collective_reduce"() ({
+  ^bb0(%arg0: tensor<f32>, %arg1: tensor<f32>):
+    %sum = stablehlo.add %arg0, %arg1 : tensor<f32>
+    "stablehlo.return"(%sum) : (tensor<f32>) -> ()
+  }) {
+    replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>,
+    has_dynamic_root
+  } : () -> ()
+  func.return
+}
```
Comment (mrguenther) [inline stablehlo/tests/ops_stablehlo.mlir:7006]: This test case brings a related question to mind: what's the expected behavior if `collective_reduce` isn't passed any operands while `has_dynamic_root=false`? We might want to add another test case for that.

Also, if the answer is that it's always an error to pass zero operands to `collective_reduce`, we might want to update this test case so that it passes exactly one argument and still expects an error because `has_dynamic_root=true`.

File: stablehlo/tests/ops_stablehlo.mlir [hunk 7]
```diff
@@ -6837,3 +6837,232 @@ func.func @all_gather_with_mesh_axes_replica_group(%operand: tensor<16x8xf32>) -
...
+  // expected-error@+1 {{The reduction-region expected to return some value(s)}}
+  %0 = "stablehlo.collective_reduce"(%operand) ({
+  ^bb0(%arg0: tensor<f32>, %arg1: tensor<f32>):
+    %sum = stablehlo.add %arg0, %arg1 : tensor<f32>
+    "stablehlo.return"() : () -> ()
+  }) {
+    replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>
+  } : (tensor<4xf32>) -> tensor<4xf32>
+  func.return %0 : tensor<4xf32>
+}
```
Comment (mrguenther) [inline stablehlo/tests/ops_stablehlo.mlir:7084]: Is the reducer region allowed to return multiple values? The `expected-error` message suggests that it's allowed; we might want a test case for it if so.

**Review by mrguenther — COMMENTED** [review]
LGTM, though I had a couple of questions about the spec in certain edge cases that might warrant additional test cases.

File: stablehlo/tests/ops_stablehlo.mlir (hunk 6 above)
Comment (Tixxx) [reply stablehlo/tests/ops_stablehlo.mlir:7006]: sounds good, i added two more tests for the two negative cases and also changed the error messages to be more explicit.

File: stablehlo/tests/ops_stablehlo.mlir (hunk 7 above)
Comment (Tixxx) [reply stablehlo/tests/ops_stablehlo.mlir:7084]: no, im implementing this based on the contract from other collective reduction ops, it only returns a single value for the reduce computation

**Comment by Tixxx** [comment]
sorry about the frequent pings but can we land this soon? this is actually blocking my xla [pr](https://github.com/openxla/xla/pull/46104). thanks!

**Comment by mrguenther** [comment]
Oh, I'm sorry, I thought I'd approved it. LGTM!

**Review by mrguenther — APPROVED** [review]

**Comment by Tixxx** [comment]
Thanks for landing this!
